src/service/position-observer-impl.js

```diff
@@ -0,0 +1,410 @@
+      this.inScroll_ = false;
+    };
+
+    this.unlisteners_.push(this.viewport_.onScroll(() => {
+      this.inScroll_ = true;
+      this.scrollTimer_ = Date.now();
+      this.schedulePass_();
+      // TODO(@zhouyx, #9208) using debounce.
+      setTimeout(stopScroll.bind(this), 500);
+    }));
+    this.unlisteners_.push(this.viewport_.onChanged(() => {
+      this.vsync_.measure(() => {
+        this.pass_(true);
+      });
+    }));
+    // QQQ: Already call this.updateEntryPosition_(entry) in observe()? Needed?
+    // initial pass
+    this.vsync_.measure(() => {
+      this.pass_(true);
+    });
+  }
+
+  /* @override */
+  stopCallback() {
+    while (this.unlisteners_.length) {
+      const unlisten = this.unlisteners_.pop();
+      unlisten();
+    }
+  }
+
+  /** @param {boolean=} recursive */
+  schedulePass_(recursive) {
+    // TODO (@zhouyx, #9208):
+    // P1: account for effective fidelity using this.effectiveFidelity
+    // P2: do passes on onDomMutation (if available using MutationObserver)
+    if (!recursive && this.measure_) {
+      // call of schedulePass_ from viewport onScroll
+      // Do nothing if currently measure with calling schedulePass recursively
+      return;
+    }
+    this.measure_ = true;
+    if (!this.inScroll_) {
+      // not in scroll, do not need to measure
+      this.measure_ = false;
+      return;
+    }
+    this.vsync_.measure(() => {
+      this.pass_();
+      this.schedulePass_(true);
+    });
+  }
+
+  /**
+   * @param {boolean=} force
+   * @private
+  */
+  pass_(force) {
+    for (let i = 0; i < this.entries_.length; i++) {
+      const entry = this.entries_[i];
+      if (!force && entry.turn != 0) {
+        // Not ready for their turn yet.
+        entry.turn--;
+        continue;
+      }
+      // Reset entry.turn value.
+      if (!force) {
+        if (entry.fidelity == PositionObserverFidelity.LOW) {
+          entry.turn = LOW_FEDELITY_FRAME_COUNT;
+        }
+      }
+
+      this.updateEntryPosition_(entry);
+    }
+  }
+
+  /** @override */
+  updateEntryPosition_(entry) {
+    // get layoutBoxes relative to doc.
+    const elementBox = layoutRectFromDomRect(
+        entry.element./*OK*/getBoundingClientRect());
+    const win = this.ampdoc_.win;
+
+    let viewportWidth = win./*OK*/innerWidth;
+    let viewportHeight = win./*OK*/innerHeight;
+    if (!viewportWidth || !viewportHeight) {
+      const el = win.document.documentElement;
+      viewportWidth = el./*OK*/clientWidth;
+      viewportHeight = el./*OK*/clientHeight;
+    }
+    const viewportBox = layoutRectLtwh(0, 0, viewportWidth, viewportHeight);
+
+    // Return { positionRect: <LayoutRectDef>, viewportRect: <LayoutRectDef>}
+    entry.trigger(/** @type {PositionInViewportEntryDef}*/ ({
+      positionRect: elementBox,
+      viewportRect: viewportBox,
+    }));
+  }
+}
+
+export class InaboxPositionObserver extends AbstractPositionObserver {
```

Comment:
Change name to make it clear what it does. Two questions to be answered: who is the host and for what purpose it exists (who's the client).